disposeOnUmount + new patch lifecycle method + updated deps

xaviergonz opened the pull request:
This pr adds a disposeOnUnmount as seen in https://github.com/mobxjs/mobx-react/issues/561

Closes #561

Also since I was getting weird interactions between `@observer` and this decorator (since both try to tweak componentWillUnmount) I had to adjust the way of patching lifecycle methods. The good thing is that I think it is more robust (should work with native classes, works with lifecycle methods set as arrow functions, is able to compose "mixins", etc)

Also updated dependencies

mweststrate commented:
I'm on holiday coming weeks (without laptop) so there will be some delay in reviewing, but I'll do my best to catch up the week after :-)

xaviergonz commented:
No problem :) Just updated dependencies as well.
also now travis shows the same console.error I get while running it in my pc (since I switched it to use yarn instead of npm)

xaviergonz commented:
fixed it, it was "for (const x of a)" transpiling into using Symbol.iterator, so just changed them to foreach

mweststrate commented on src/disposeOnUnmount.js:
```diff
@@ -0,0 +1,53 @@
+
+    // add property key / function we want run (disposed) to the store
+    const componentWasAlreadyModified = !!target[storeKey]
+    target[storeKey] = target[storeKey] || []
```
NIT: simplify these 2 lines to: `const store = target[storeKey] || (target[storeKey] = [])`

xaviergonz replied:
will do

mweststrate commented on src/index.d.ts:
```diff
@@ -58,6 +58,15 @@ export function inject<S, P, I, C>(
+/**
+ * disposeOnUnmount
+ */
+export function disposeOnUnmount(target: React.Component, propertyKey: string): void
```
The official type for a field decorator is: `(target: Object, property: PropertyKey, descriptor?: PropertyDescriptor): void`

xaviergonz replied:
I can change the others, but should the target be changed? as it is right now it prevents the decorator from being used on non react component classes

xaviergonz replied:
Just took a look at lib.es5.d.ts and it shows
`declare type PropertyDecorator = (target: Object, propertyKey: string | symbol) => void;`
since the decorator should only work over react component classes and it doesn't work over symbol properties, shouldn't I leave it as is?

mweststrate commented on src/index.d.ts:
```diff
@@ -58,6 +58,15 @@ export function inject<S, P, I, C>(
+ * disposeOnUnmount
+ */
+export function disposeOnUnmount(target: React.Component, propertyKey: string): void
+export function disposeOnUnmount<TF extends Function | Function[]>(
```
I would explicitly type `Function` as `Disposer`, where `Disposer` is `() => void`

xaviergonz replied:
will do

mweststrate commented on src/index.js:
```diff
@@ -21,6 +21,7 @@ export {
 
 export { default as Provider } from "./Provider"
 export { default as inject } from "./inject"
+export { disposeOnUnmount } from "./disposeOnUnmount"
```
For naming, the name is clear, but quite long. Would there be a downside to calling it `onUnmount`?

mweststrate replied:
Not entirely sure though

xaviergonz replied:
I'm open for suggestions, so up to you 😁

xaviergonz replied:
I'm open for suggestions, so up to you 😁

xaviergonz replied:
I'm just weary of how it'd look on the non decorator version
onUnmount(reaction...)
makes it sound like the reaction will trigger on unmount or something

mweststrate replied:
Ok, let's keep it as is! If it is to long people can always alias during import :)

mweststrate commented on src/utils/utils.js:
```diff
@@ -3,3 +3,59 @@ export function isStateless(component) {
+    return methodMixins
+}
+
+export function patch(target, methodName, mixinMethod, runMixinFirst = false) {
```
could you elaborate why this enhanced patch was needed?

xaviergonz replied:
for two reasons. 
the first one is that the patching method would only support one decorator override, so when this decorator was being used with observer one of the two would run, but not both.
the second one is that the current patch method works OK for prototype methods, but not for property methods defined as arrow function, so this new patch method supports both

xaviergonz replied:
Quoting the PR description:
Also since I was getting weird interactions between @observer and this decorator (since both try to tweak componentWillUnmount) I had to adjust the way of patching lifecycle methods. The good thing is that I think it is more robust (should work with native classes, works with lifecycle methods set as arrow functions, is able to compose "mixins", etc)

mweststrate commented:
@xaviergonz thanks! Great PR :-). Left some questions

xaviergonz commented:
code review changes done, except for:
* the decorator typing, see comments I added about it
* name still not changed, (see comment I added about it)

xaviergonz commented:
so is all good now?

mweststrate commented:
Will try to review the rest later tonight!

Op do 11 okt. 2018 om 20:50 schreef Javier Gonzalez <
notifications@github.com>:

> so is all good now?
>
> —
> You are receiving this because you were mentioned.
>
> Reply to this email directly, view it on GitHub
> <https://github.com/mobxjs/mobx-react/pull/563#issuecomment-429076507>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABvGhFJVBVdPYtsX7r_48NuGa32_lHIiks5uj5MOgaJpZM4W_sC2>
> .
>

mweststrate commented:
(currently at meetup)

Op do 11 okt. 2018 om 21:03 schreef Michel Weststrate <mweststrate@gmail.com
>:

> Will try to review the rest later tonight!
>
> Op do 11 okt. 2018 om 20:50 schreef Javier Gonzalez <
> notifications@github.com>:
>
>> so is all good now?
>>
>> —
>> You are receiving this because you were mentioned.
>>
>> Reply to this email directly, view it on GitHub
>> <https://github.com/mobxjs/mobx-react/pull/563#issuecomment-429076507>,
>> or mute the thread
>> <https://github.com/notifications/unsubscribe-auth/ABvGhFJVBVdPYtsX7r_48NuGa32_lHIiks5uj5MOgaJpZM4W_sC2>
>> .
>>
>

xaviergonz commented:
no need to hurry :) Btw, I just created a require('@mweststrate') tag on MST for issues that cannot advance until you make a decision, just so you can quickly filter over issues that absolutely require you (eventually) :)

mweststrate commented on src/disposeOnUnmount.js:
```diff
@@ -0,0 +1,52 @@
+const storeKey = "__$mobxDisposeOnUnmount"
+
+function customComponentWillUnmount() {
+    this[storeKey].forEach(propKeyOrFunction => {
```
NIT: `this[storeKey].splice(0).forEach` prevents concurrent modifications, and doesn't need the reset of line 19. (No need to change it if there are no other changes for this MR, just a nice pattern in general, imho)

xaviergonz replied:
OK, actually forEach does, but (for of) doesn't :)
still, since this once uses forEach it should be ok

xaviergonz commented on src/disposeOnUnmount.js:
```diff
@@ -0,0 +1,52 @@
+
+const storeKey = "__$mobxDisposeOnUnmount"
+
+function customComponentWillUnmount() {
```
sounds good too :)

mweststrate commented on src/utils/utils.js:
```diff
@@ -3,3 +3,59 @@ export function isStateless(component) {
+    }
+    getFunction["__$mobxMixin"] = true
+
+    const newDefinition = {
```
Optimization: performance / memory consumption is significant lower if property descriptors are cached and reused per name. To achieve that `getFunction` needs to be generic, for that, `realMethod` needs not to be used. That could be done by for example storing the real method on the `mixins.real`  property

mweststrate commented on src/utils/utils.js:
```diff
@@ -3,3 +3,59 @@ export function isStateless(component) {
+        set: value => {
+            realMethod = value
+        },
+        configurable: false
```
keep this one on `true`. Many testing tools (if people want to stub) or things like HMR trip over non-configurable properties

xaviergonz replied:
ok, hopefully those tools won't break it badly :)

mweststrate commented on test/ts/compile-ts.tsx:
```diff
@@ -232,14 +270,34 @@ class InjectSomeStores extends Component<{ x: any }, {}> {
-	<Injected.wrappedComponent x={3} />
-}
\ No newline at end of file
+    // just to make sure it compiles
```
It would be great if we had an actual Typescript test for this, as the decorator implementation of TS is different from the babel one, but it might be too much of a hassle to add it to the current setup, I am confident enough that this case is simple enough to not fail (famous last words)

mweststrate commented on src/utils/utils.js:
```diff
@@ -3,3 +3,59 @@ export function isStateless(component) {
 }
+
+function getMixins(target, methodName) {
+    const mixins = (target["__$mobxMixins"] = target["__$mobxMixins"] || {})
```
NIT: constants (and Symbols)

xaviergonz replied:
sure

mweststrate commented:
@xaviergonz Ok, I think I finished my review :-). Only reservation is that I think double-patching the old way should work fine, and has less risk of conflicting with other tools downstream. On the other hand, fixing the subtle arrow func issue nice, as people run into that a few times. So let's proceed with this approach unless proven otherwise. I think it deserves to be a minor for that release at least though :-).

xaviergonz commented:
code review changes done, while at it i also changed some other internal props to symbols

xaviergonz commented:
ah forgot the Optimization: performance / memory consumption one

xaviergonz commented:
definition optimization done as well

xaviergonz commented:
There's one thing that worries me about the definition optimization though.
If two different classes use the same prop def for the same method and a tool decides only to patch up one of the components, will it end up patching both?
If so, should the definition be cached per class as well?

mweststrate commented:
@xaviergonz if a tools patches something, the descriptor gets replaced, so the descriptors themselves are practically immutable, making this save (afaik grabbing a property descriptor using reflection, and then changing it has no effect)

xaviergonz commented:
while at it also improved a bit the symbol polyfill (it was based on math.random, which seemed flaky) and unified them all over the utils function newSymbol

mweststrate commented:
Ok, will try to merge tomorrow, some things I want to check locally, so hope I'll find tomorrow time to wrap up :)
